### PR TITLE
add distributed example for CUDA RPC

### LINKED ISSUE #46321: Add distributed examples into PyTorch CI tests
### Targeted Examples

This will help us to capture tutorial/examples failures caused by BC-breaking changes. So the test should ideally be the same as the examples. However, as some examples depends on non-PyTorch libraries, we will need to replace those with some dummy implementation. Below is a list of existing examples.

- [ ] [Distributed Reinforcement Learning using RPC and RRef](https://pytorch.org/tutorials/intermediate/rpc_tutorial.html#distributed-reinforcement-learning-using-rpc-and-rref). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/rl). One caveat is that, since we cannot let PyTorch depend on `gym`, we will need to create a dummy environment. This is fine, as we are only testing API correctness. It should be sufficient to return random states and let the training loop to run a small number of iterations.
- [ ] [Distributed RNN using Distributed Autograd and Distributed Optimizer](https://pytorch.org/tutorials/intermediate/rpc_tutorial.html#distributed-rnn-using-distributed-autograd-and-distributed-optimizer). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/rnn).
- [ ] [Implementing a Parameter Server using Distributed RPC Framework](https://pytorch.org/tutorials/intermediate/rpc_param_server_tutorial.html). Example source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/parameter_server).
- [ ] [Batch-Updating Parameter Server](https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-updating-parameter-server). Source code is [here](https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py). Similar to the first item, we cannot make PyTorch depend on torchvision, so please choose some dummy models instead.
- [ ] [Batch-Processing CartPole Solver](https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-processing-cartpole-solver). Source code is [here](https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/reinforce.py). The test cannot depend on `gym`, so please use some dummy environments.
- [ ] [Distributed Pipeline Parallelism Using RPC](https://pytorch.org/tutorials/intermediate/dist_pipeline_parallel_tutorial.html). Source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/pipeline). Again, cannot depend on `torchvision`.
- [ ] [Combining DDP with RPC](https://pytorch.org/tutorials/advanced/rpc_ddp_tutorial.html). Source code is [here](https://github.com/pytorch/examples/tree/master/distributed/rpc/ddp_rpc).

### Implementation

TBA

### PR DESCRIPTION
**Pull request by gcramer23**
work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

Differential Revision: [D27420409](https://our.internmc.facebook.com/intern/diff/D27420409)

### CODE REVIEW

**Comment by pritamdamania87** [comment]
What is the purpose of this new test? This feature is already tested here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py.

File: torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+)
+from torch.testing._internal.dist_utils import (
+    dist_init,
+)
+
+NUM_EMBEDDINGS = 100
+EMBEDDING_DIM = 16
+
+
+class HybridModel(torch.nn.Module):
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py:27]: Hey @gcramer23, there might be some misunderstanding, the two onboarding tasks mentioned in the onboarding doc are the following. Did you get this task from someone else?

1. Add distributed example for CUDA RPC (https://pytorch.org/tutorials/recipes/cuda_rpc.html) to existing PyTorch CI tests.

* https://github.com/pytorch/pytorch/issues/46321 (CUDA RPC)

2. Add a tests to test_c10d.py to explicitly cover DDP on mixed dense-sparse models. Currently we have a sparse-only module tests.

* https://github.com/pytorch/pytorch/issues/54445


The second task should only use DDP (gloo backend) for mixed dense and sparse tensors.

Comment (gcramer23) [reply torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py:27]: @mrshenli 
My mistake must have had a misunderstanding when looking at this issue https://github.com/pytorch/pytorch/issues/46321. I will add a distributed example for the CUDA RPC. To clarify that is the reserved task in the targeted examples?

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        return 2
+
+    @dist_init(setup_rpc=False)
+    def test_cuda_distributed_rpc(self):
+
+        if not torch.cuda.is_available():
+            sys.exit(TEST_SKIPS['no_cuda'].exit_code)
+
+        if torch.cuda.is_available() and torch.cuda.device_count() < 2:
+            sys.exit(TEST_SKIPS['multi-gpu'].exit_code)
```
Comment (H-Huang) [inline torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:70]: We can use the `skip_if_not_multigpu` and `skip_if_no_gpu` decorators from  `torch.testing._internal.common_distributed` instead

File: torch/testing/_internal/distributed/rpc_utils.py [hunk 3]
```diff
@@ -48,6 +48,7 @@
 from torch.testing._internal.distributed.rpc.examples.reinforcement_learning_rpc_test import (
     ReinforcementLearningRpcTest,
 )
+from torch.testing._internal.distributed.rpc.cuda_distributed_rpc_test import CudaDistributedRPCTest
```
Comment (H-Huang) [inline torch/testing/_internal/distributed/rpc_utils.py:51]: Can we put this file under `torch.testing._internal.distributed.rpc.examples` folder instead?

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import os
```
Comment (H-Huang) [inline torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:1]: Add comment with link to example on the site

File: torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py (hunk 1 above)
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py]: no worries. Yep, it's the reserved task, which links to this page: https://pytorch.org/tutorials/recipes/cuda_rpc.html

Let me add your name to it.

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py (hunk 2 above)
Comment (H-Huang) [reply torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:70]: Looks like there is something wrong with the `skip_if_not_multigpu` decorator implementation as it is currently skipping all tests. Thanks @gcramer23 for bringing this up!

Comment (gcramer23) [reply torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:70]: the issue I created https://github.com/pytorch/pytorch/issues/54887

File: torch/testing/_internal/distributed/rpc_utils.py (hunk 3 above)
Comment (gcramer23) [reply torch/testing/_internal/distributed/rpc_utils.py]: okay moved it.

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py (hunk 4 above)
Comment (gcramer23) [reply torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py]: okay added comment with example link.

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py (hunk 2 above)
Comment (H-Huang) [reply torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py]: As Shen mentioned, adding `@skip_if_lt_x_gpu(2)` and `@skip_if_no_gpu` should work instead

**Review by H-Huang — APPROVED** [review]
Also use `ghimport -s https://github.com/pytorch/pytorch/pull/54678` on your devserver to import this PR to phabricator.

File: torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py (hunk 2 above)
Comment (gcramer23) [reply torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py]: added

**Review by mrshenli — COMMENTED** [review]
Is it possible to add some `self.assert*` to make sure we have parity between CPU, CUDA, and local fwd/bwd?

**Comment by mrshenli** [comment]
> do you mean self.assert on the cpu and cuda loss and gradients?

Hey @gcramer23, yep, I was referring to that. Could you please confirm after you set random seeds, model weights and input tensors have same values between CPU and CUDA runs? If so, it might relates to this issue: https://github.com/pytorch/pytorch/issues/54771

Let's hold on a bit and try again when that issue is fixed.

**Comment by gcramer23** [comment]
>  Could you please confirm after you set random seeds, model weights and input tensors have same values between CPU and CUDA runs?

yes the model weights and input tensors have the same values between CPU and CUDA runs.
